### PR TITLE
Some guidance on writing unit tests

### PR DESCRIPTION
**Pull request by gilles-peskine-arm**
* Utility assert macros
* Buffer management
* PSA init/deinit
* Document the data
* Where to find test data    

A lot of this document is out of date. I am not volunteering to bring it up to date at this time. My update only covers topics that didn't apply, or that we weren't thinking of, back when this document was written.

### CODE REVIEW

File: kb/development/test_suites.md [hunk 1]
```diff
@@ -114,3 +114,112 @@ void test_function_example( char *input, char *expected_output, int expected_ret
...
+    * In rare cases where a part of the test code shouldn't be reached, the convention is to use `TEST_ASSERT(!"explanation of why this shouldn't be reached")`.
+
+### Buffer allocation
+
+When a function expects an input or an output to have a certain size, you should pass it an allocated buffer with exactly the expected size. The continuous integration system runs tests in many configurations with Asan or Valgrind, and these will cause test failures if there is a buffer overflow or underflow.
+
+For output buffers, it's usually desirable to also check that the function works if it's given a buffer that's larger than necessary, and that it returns the expected error if given a buffer that's too small.
+
+Here is an example of a test function that checks that a library function has the desired output for a given input.
+```
```
Comment (tom-cosgrove-arm) [inline kb/development/test_suites.md:140]:
```suggestion
```c
```

File: kb/development/test_suites.md [hunk 2]
```diff
@@ -114,3 +114,112 @@ void test_function_example( char *input, char *expected_output, int expected_ret
...
+Test data for cryptographic algorithms should be, in order of preference:
+
+* Official test vectors taken from a standards document.
+* Generated by a reference implementation from the authors of the algorithm.
+* Generated by a widespread implementation such as OpenSSL or Cryptodome.
+* Generated by an independent implementation.
+* Obtained through manual means, possibly by patching together bits of other tests. For example, test output multipart operation function can be obtained by splitting the output of a one-shot operation function. Edge cases for parsing can be constructed by manually tweaking nominal cases.
+* As a last resort, obtained by running the library once. This is a last resort since it cannot validate that the implementation is correct, it can only protect against future behavior changes. This should only be done when there is no other way, for example to construct a non-regression test in an edge case if we're very confident from code review that our bug fix is correct.
+
+Always [document how the test data was constructed](#document-the-test-data).
```
Comment (tom-cosgrove-arm) [inline kb/development/test_suites.md:225]: Is it necessary to repeat this given it's actually a section heading less than a page above? Not sure how it counts as "interoperability testing"

Comment (yanesca) [reply kb/development/test_suites.md:225]: I think it is useful to emphasise that it needs to be documented and recording the source of the test vectors is even more important here.

**Review by yanesca — DISMISSED** [review]
LGTM.

File: kb/development/test_suites.md (hunk 2 above)
Comment (gilles-peskine-arm) [reply kb/development/test_suites.md]: I meant “whichever of these options you choose, remember to write down what you did”. I've reworded the sentence to actually say that.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

File: kb/development/test_suites.md [hunk 3]
```diff
@@ -196,3 +196,30 @@ In a test case that always uses PSA crypto, call `PSA_INIT()` at the beginning a
...
+
+When adding a new file in `tests/data_files`, if possible, do it by editing `tests/data_files/Makefile` to add executable instructions for creating those files, and then run `make` and commit the result. Ideally removing the files and re-running `make` should produce identical files, however in many cases this is not practical because the generation is randomized or depends on the current time (for certificates and related data), and we accept that.
+
+### Interoperability testing
+
+Test data for cryptographic algorithms should be, in order of preference:
+
+* Official test vectors taken from a standards document.
+* Generated by a reference implementation from the authors of the algorithm.
+* Generated by a widespread implementation such as OpenSSL or Cryptodome.
```
Comment (gilles-peskine-arm) [inline kb/development/test_suites.md]: On second thoughts I don't think we should systematically privilege implementations from the authors. Yes, they're more official, but they're also often harder to use. So I don't want to discourage using a popular implementation like OpenSSL. Obviously if we suspect an OpenSSL bug we can dig down further — and even so we might privilege compatibility with the de facto standard over the de jure standard.

**Review by tom-cosgrove-arm — APPROVED** [review]

**Review by yanesca — APPROVED** [review]
LGTM

**Comment by AndrzejKurek** [comment]
I can see that we don't have an "approved" label on this repository, so I'll just write it here instead for visibility: this PR looks like it's good to merge! :)
